### PR TITLE
refactor: AssetRecyclerList

### PR DESCRIPTION
**Pull request by cawfree**
Replaces the `AssetRecyclerList` with a functional component equivalent.

  - Removes the `getStableId` function.
    - The reason `getStableId` was removed is because it does not work synchronously with render props; an older instance of the `DataProvider` can be temporarily retained and generate inconsistent property keys.
      - Previously, the `RecyclerAssetList` relied upon a circular stateful reference to achieve this. 
    - My intention is to rely on `recyclerlistview`'s internal list management, which renders consistently compared to the previous implementation.
  - Replace `hasRowChanged` function with `isEqual` from `react-fast-compare`.
    - This removes layout-specific change detection, which increased component complexity and relied upon a priori knowledge of element structure to perform element-specific property comparison. Here, we defer this responsibility to `react-fast-compare`.
  -  Removes global render variables.
  - Added typing information to local data models.

### CODE REVIEW

File: src/components/asset-list/RecyclerAssetList/index.tsx
Comment (osdnk) [inline src/components/asset-list/RecyclerAssetList/index.tsx:732]: We have hooks for all of those HOCs

File: src/components/asset-list/RecyclerViewTypes.js
Comment (osdnk) [inline src/components/asset-list/RecyclerViewTypes.js:42]: ?

File: src/components/asset-list/RecyclerAssetList/index.tsx
Comment (osdnk) [inline src/components/asset-list/RecyclerAssetList/index.tsx:1]: Can you remove the previous implementation?

File: src/components/asset-list/RecyclerAssetList/index.tsx
Comment (osdnk) [inline src/components/asset-list/RecyclerAssetList/index.tsx:315]:
```suggestion
              renderItem: section.renderItem,
```

File: src/components/asset-list/RecyclerAssetList/index.tsx
Comment (osdnk) [inline src/components/asset-list/RecyclerAssetList/index.tsx:307]:
```suggestion
                renderItem: section.renderItem,
```

**Review by osdnk — APPROVED** [review]
Looks good. I think we can merge this in the current state. Each of my comments is nonblocking. Feel free to merge

**Comment by osdnk** [comment]
Ofc, assuming it works

**Comment by brunobar79** [comment]
I think @skylarbarrera mentioned there were some minor bugs, so let's wait for his feedback before merging

**Comment by skylarbarrera** [comment]
i like to use 0xb1 to really push the boundaries  & performance is alot better 🙏  

Going to build this to my phone and use it for the day to make sure its 👌 , i think it MAY slow down the app after using for awhile need to confirm 

Sticky Header needs some help, its pretty jumpy on the initial jump, performs better on android 🤔
http://cloud.skylarbarrera.com/Screen-Recording-2021-03-11-09-57-50.mp4

**Comment by cawfree** [comment]
@skylarbarrera Awesome! I'll start working on these fixes, thanks for the thorough check!

File: src/components/asset-list/RecyclerAssetList/index.tsx
Comment (cawfree) [reply src/components/asset-list/RecyclerAssetList/index.tsx:1]: I was able to remove `findIndex` and `isNil`, but I'm hesitant to remove `get` because this is a complex operation and is used throughout the codebase.

It looks like [refactoring `get` should be easy](https://youmightnotneed.com/lodash/), unless there's already an existing implementation in the codebase that I should import instead?

**Comment by cawfree** [comment]
@skylarbarrera 

`ScrollView` slowness on iOS might down to [this](https://github.com/facebook/react-native/blob/1c564c3a1f28e780fe0bf33871f61e17730fd512/Libraries/Components/ScrollView/ScrollView.js#L1167). But I'm a little sceptical about this since it worked fine without `scrollEventThrottle` in the past, but it would explain the performance difference between Android and iOS. 

I'll spend a little more time looking at what `onScroll` is doing to determine if we've accidentally increased the overhead of this call.

**Comment by brunobar79** [comment]
> @skylarbarrera 
> 
> 
> 
> `ScrollView` slowness on iOS might down to [this](https://github.com/facebook/react-native/blob/1c564c3a1f28e780fe0bf33871f61e17730fd512/Libraries/Components/ScrollView/ScrollView.js#L1167). But I'm a little sceptical about this since it worked fine without `scrollEventThrottle` in the past, but it would explain the performance difference between Android and iOS. 
> 
> 
> 
> I'll spend a little more time looking at what `onScroll` is doing to determine if we've accidentally increased the overhead of this call.

Feel free to make a different implementation for each OS if it helps!

**Comment by skylarbarrera** [comment]
@cawfree sticky header is 👌  now! I do see the following on Android, but other than that lgtm http://cloud.skylarbarrera.com/Screen-Recording-2021-03-12-12-05-50.mp4

**Comment by cawfree** [comment]
Okay, the `StickyHeader` didn't like being dynamically populated during rendering. My intuition is that this was triggering `onLayout` and confusing the scroll offset.

I've moved the `CoinDivider` to a dedicated view which we modulate `opacity` and `pointerEvents` using `setNativeProps`. This way we don't have to cause a state change to achieve the sticky effect, and subsequently the glitchy behaviour goes away!

**Review by skylarbarrera — APPROVED** [review]
🐐

**Comment by brunobar79** [comment]
LFG!!!

**Pull request merged**
